### PR TITLE
[usage] Improve pagination behavior

### LINKED ISSUE #11530: Improve pagination layout, design, and structure
### Problem to solve

Following up from https://github.com/gitpod-io/gitpod/pull/11481, we need to update the pagination component. Some UX issues currently involved include:
1. Lack of proper disabled button states, see Previous and Next at the first or last page.
2. Lack of proper default, hover, and active states on buttons.
3. Lack of truncation when pages are more than what we can show.
4. ...

Screenshot 2022-07-21 at 3 20 44 PM

### PR DESCRIPTION
**Pull request by laushinka**
## Description
Improves the pagination behavior.
Note: This has a /hold label because it has a test commit that needs to be removed later.

| 1 | 2 | 3 |
| - | - | - |
| Screenshot 2022-08-03 at 16 26 44 | Screenshot 2022-08-03 at 16 26 54 | Screenshot 2022-08-03 at 16 27 01 |

## Related Issue(s)

Fixes #11530 

## How to test
This PR includes a file `dummy.ts` so that the pagination behavior when there are too many pages can be tested.
The number of results per page is also made smaller.
1. Join [this team](https://lau-pagina7108ace06a.preview.gitpod-dev.com/teams/join?inviteId=ccd62582-06a3-4fd3-928c-ad8159670a30).
2. Go to https://lau-pagina7108ace06a.preview.gitpod-dev.com/t/gitpod/usage.
3. Play around with the pagination.

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@laushinka: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by laushinka** [comment]
Due to a [test commit](https://github.com/gitpod-io/gitpod/pull/11800/commits/5aabd213ffff3f1a909b9d057fb48268f7b89e6d):

/hold

**Comment by andrew-farries** [comment]
Are you still able to access the preview for this? It doesn't work for me:

image

**Comment by laushinka** [comment]
> Are you still able to access the preview for this? It doesn't work for me:

Let me re-run it - that usually fixes it.

/werft run

**Comment by laushinka** [comment]
@andrew-farries All good now https://lau-pagina7108ace06a.preview.gitpod-dev.com/t/gitpod/usage

File: components/dashboard/src/teams/TeamUsage.tsx
Comment (laushinka) [inline components/dashboard/src/teams/TeamUsage.tsx:35]: I removed the dummy file and instead set this to 1 per page. Will change to `15` before merging.

**Comment by andrew-farries** [comment]
Can you make me an owner of the team? Otherwise: 

image

File: components/dashboard/src/custom-hooks/usePagination.ts
Comment (andrew-farries) [inline components/dashboard/src/custom-hooks/usePagination.ts:9]: I'm not sure why this needs to be a hook. AIUI this returns an array of page numbers based on its arguments. I think the `calculatedPagination` function could be inlined into the `Pagination` component and called there directly.

Comment (laushinka) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: The main reason in extracting it was for readability and ease of maintenance. The usePagination logic is quite complex and it would be much harder to maintain with other noise around it.

Comment (andrew-farries) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: completely agree to make it a separate function rather than doing it inline in the component. I just wonder if it can be an ordinary function rather than a hook.

Comment (laushinka) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: Aha. I think technically yes, but I did want to make use of the `useMemo` for better performance, and therefore it has to be either a Hook or a React component. Do you think it would be overhead to maintain?

Comment (andrew-farries) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: I'm thinking of converting the hook to just a function called `getPageNumbers` (or whatever), that takes the same arguments as  the hook does now and returns an array of page numbers (as the hook does now).

If the function is created outside the `Pagination` component, it won't need to be memoized because its identity won't change between renders of the `Pagination` component.

Comment (laushinka) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: Hmm, makes sense. The new function will then have to be called inside a useEffect, right? Whenever each of the args change. Also let me know if a quick sync would be faster 😬

Comment (andrew-farries) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: I think there is no need for a `useEffect` - the function is just a helper for turning current `props` into an array of page numbers - no synchronization with anything external that would need a `useEffect`.

essentially the pattern is:

```ts
const func getPageNumbersFromProps(...) => { return [1,2,3,4] }

const MyComponent = (props) => {
  const xs = getPageNumbersFromProps(props)

  return (<ul>
    xs.map(x => <li>x</li>
  </ul>)
}
```

Comment (laushinka) [reply components/dashboard/src/custom-hooks/usePagination.ts:9]: Ah sorry not sure why I said useEffect I meant useState for updating the values, but they do get updated in the parent component. Will give this a shot after lunch.

File: components/dashboard/src/Pagination/Pagination.tsx
Comment (andrew-farries) [inline components/dashboard/src/Pagination/Pagination.tsx:14]: nit: I think you can give this a proper type; hovering over a setter should give you the type.

**Review by andrew-farries — APPROVED** [review]
Nice. I like this more with the unit test and without the hook 👍 

/hold for the typing issue and this comment: https://github.com/gitpod-io/gitpod/pull/11800/files#r936443747

File: components/dashboard/src/Pagination/Pagination.tsx
Comment (laushinka) [reply components/dashboard/src/Pagination/Pagination.tsx]: Lol totally forgot about this one. Done!

Comment (laushinka) [reply components/dashboard/src/Pagination/Pagination.tsx]: Btw it's also on hold because I still need to change the `resultsPerPage` that I changed for this PR.

Comment (laushinka) [reply components/dashboard/src/Pagination/Pagination.tsx]: > yeah, I linked your comment in the review summary comment.

My brain is so not working right now 🤣

File: components/dashboard/src/teams/TeamUsage.tsx
Comment (laushinka) [reply components/dashboard/src/teams/TeamUsage.tsx]: Already put back and will unhold.

**Comment by laushinka** [comment]
Thank you @andrew-farries for reviewing and making sure I don't over-engineer 🧡

**Pull request merged**
